File: entities/inventory.go
package entities

import (
	"fmt"
	"log"
	"math/rand"

	"legend-of-adventure/events"
)

type InventoryOwner interface {
	UpdateInventory()
	SetEffect(effect string, ttl int)
	ID() string
	Receive() chan<- *events.Event
}

type Inventory struct {
	Owner    InventoryOwner
	inv      []string
	counts   []int
	capacity int
}

func NewInventory(owner InventoryOwner, capacity int) *Inventory {
	slots := make([]string, capacity)
	counts := make([]int, capacity)
	return &Inventory{owner, slots, counts, capacity}
}

func (self *Inventory) Get(index int) (string, int) {
	return self.inv[index], self.counts[index]
}

func (self *Inventory) Capacity() int {
	return self.capacity
}

func (self *Inventory) NumItems() int {
	count := 0
	for _, v := range self.inv {
		if v != "" {
			count = count + 1
		}
	}
	return count
}

func (self *Inventory) IsFull() bool {
	for _, v := range self.inv {
		if v == "" {
			return false
		}
	}
	return true
}

func (self *Inventory) Give(item string) (bool, int) { // Success, Slot
	for i, v := range self.inv {
		if v == item && self.counts[i] < INV_MAX_STACK {
			self.counts[i]++
			self.Owner.UpdateInventory()
			return true, i
		}
	}

	if self.IsFull() {
		return false, 0
	}

	for i, v := range self.inv {
		if v == "" {
			self.inv[i] = item
			self.counts[i] = 1
			self.Owner.UpdateInventory()
			return true, i
		}
	}
	return false, 0
}

func (self *Inventory) Clear() {
	for i := 0; i < self.capacity; i++ {
		self.ClearSlot(i)
	}
}

func (self *Inventory) ClearSlot(index int) {
	self.inv[index] = ""
	self.counts[index] = 0
}

func (self *Inventory) Consolidate() {
	for i := 0; i < self.capacity-1; i++ {
		if self.inv[i] != "" {
			continue
		}
		for j := i + 1; j < self.capacity; j++ {
			if self.inv[j] != "" {
				self.inv[i] = self.inv[j]
				self.counts[i] = self.counts[j]
				self.inv[j] = ""
				self.counts[j] = 0
				break
			}
		}
		if self.inv[i] == "" {
			break
		}
	}
}

func (self *Inventory) Cycle(command string) {
	self.Consolidate()
	count := self.NumItems()

	if command == "b" {
		first, firstCount := self.inv[0], self.counts[0]
		for i := 0; i <= count-2; i++ {
			self.inv[i] = self.inv[i+1]
			self.counts[i] = self.counts[i+1]
		}
		self.inv[count-1], self.counts[count-1] = first, firstCount
	} else {
		last, lastCount := self.inv[count-1], self.counts[count-1]
		for i := count - 1; i > 0; i-- {
			self.inv[i] = self.inv[i-1]
			self.counts[i] = self.counts[i-1]
		}
		self.inv[0], self.counts[0] = last, lastCount
	}
	self.Owner.UpdateInventory()
}

func (self *Inventory) Use(index uint, holder Animat) {
	if int(index) > self.capacity || self.inv[index] == "" {
		return
	}

	log.Println(holder.ID() + " using " + self.inv[index])

	x, y := UnpackCoords(<-(holder.Position()))

	switch self.inv[index][0] {
	case 'f':
		// Don't let the player waste the food.
		if holder.IsAtMaxHealth() {
			return
		}
		holder.IncrementHealth(FOOD_HEALTH_INCREASE)
		self.Remove(index)

		holder.Receive() <- holder.Location().GetEvent(
			events.PARTICLE_MACRO, "0.5 -0.5 eatfood 10 local", holder,
		)

		holder.Location().Broadcast(
			holder.Location().GetEvent(
				events.PARTICLE_MACRO,
				fmt.Sprintf("0.5 -0.5 eatfood 10 %s", holder.ID()),
				holder,
			),
		)

	case 'w':
		holder.Location().Broadcast(
			holder.Location().GetEvent(
				events.DIRECT_ATTACK,
				fmt.Sprintf("%f %f %s", x, y, self.inv[index]),
				holder,
			),
		)

	case 'p':
		self.Remove(index)

		effectN := rand.Intn(4)
		var effect string
		switch effectN {
		case 0:
			effect = "flip"
		case 1:
			effect = "invincible"
		case 2:
			effect = "blindness"
		case 3:
			effect = "drained"
		}
		holder.SetEffect(effect, rand.Intn(10)+10)

		holder.Location().Broadcast(
			holder.Location().GetEvent(
				events.SOUND,
				fmt.Sprintf("potion%d:%f:%f", rand.Intn(2), x, y),
				nil,
			),
		)
	}

}

func (self *Inventory) Remove(index uint) {
	self.counts[index]--
	if self.counts[index] == 0 {
		self.inv[index] = ""
	}
	self.Consolidate()
	self.Owner.UpdateInventory()
}

func (self *Inventory) Drop(dropper EntityThatCanThrow) {
	if self.inv[0] == "" {
		return
	}

	log.Println(dropper.ID() + " dropping " + self.inv[0])

	reg := dropper.Location()
	item := NewItemEntity(self.inv[0], dropper)
	self.counts[0]--
	if self.counts[0] == 0 {
		self.inv[0] = ""
	}
	reg.AddEntity(item)
	self.Consolidate()
	self.Owner.UpdateInventory()

}

File: src/inventory.js
const events = require('./events');
const ItemEntity = require('./entities/itemEntity');


const FOOD_HEALTH_INCREASE = 75;
const INV_MAX_STACK = 32;


module.exports = class Inventory {
  constructor(owner, capacity) {
    this.owner = owner;
    this.capacity = capacity;

    this.slots = new Array(capacity);
    this.counts = new Array(capacity);
    for (let i = 0; i < capacity; i++) {
      this.slots[i] = null;
      this.counts[i] = 0;
    }
  }

  numItems() {
    let sum = 0;
    for (let i = 0; i < this.capacity; i++) {
      if (!this.slots[i]) {
        continue;
      }
      sum += this.counts[i];
    }
    return sum;
  }

  isFull() {
    return this.slots.every(x => x);
  }

  give(item) {
    for (let i = 0; i < this.capacity; i++) {
      if (this.slots[i] === item && this.counts[i] < INV_MAX_STACK) {
        this.counts[i] += 1;
        this.owner.updateInventory();
        return [true, i];
      }
    }

    if (this.isFull()) {
      return [false, null];
    }

    for (let i = 0; i < this.capacity; i++) {
      if (this.slots[i]) {
        continue;
      }
      this.slots[i] = item;
      this.counts[i] = 1;
      this.owner.updateInventory();
      return [true, i];
    }

    return [false, null];
  }

  clear() {
    for (let i = 0; i < this.capacity; i++) {
      this.clearSlot(i);
    }
  }

  clearSlot(i) {
    this.slots[i] = null;
    this.counts[i] = 0;
  }

  consolidate() {
    for (let i = 0; i < this.capacity - 1; i++) {
      if (this.slots[i]) {
        continue;
      }

      for (let j = i; j < this.capacity; j++) {
        if (!this.slots[j]) {
          continue;
        }

        this.slots[i] = this.slots[j];
        this.counts[i] = this.counts[j];
        this.clearSlot(i);
      }
      if (!this.slots[i]) {
        break;
      }
    }
  }

  cycle(command) {
    this.consolidate();
    const count = this.numItems();

    if (command === 'b') {
      const firstItem = this.slots[0];
      const firstCount = this.counts[0];
      for (let i = 0; i < count - 2; i++) {
        this.slots[i] = this.slots[i + 1];
        this.counts[i] = this.counts[i + 1];
      }
      this.slots[count - 1] = firstItem;
      this.counts[count - 1] = firstCount;
    } else {
      const lastItem = this.slots[count - 1];
      const lastCount = this.counts[count - 1];
      for (let i = count - 1; i > 0; i--) {
        this.slots[i] = this.slots[i - 1];
        this.counts[i] = this.counts[i - 1];
      }
      this.slots[0] = lastItem;
      this.counts[0] = lastCount;
    }
    this.owner.updateInventory();
  }

  use(i, holder) {
    if (i >= this.capacity || !this.slots[i]) {
      return;
    }

    console.log(`${holder} using ${this.slots[i]}`);

    const x = holder.x;
    const y = holder.y;

    switch (this.slots[i][0]) {
      case 'f':
        if (holder.isAtMaxHealth()) {
          return;
        }
        holder.incrementHealth(FOOD_HEALTH_INCREASE);
        this.remove(i);

        holder.handleEvent(
          holder.region.onEvent(events.PARTICLE_MACRO),
          '0.5 -0.5 eatfood 10 local',
          holder
        );
        holder.region.broadcast(
          new events.Event(
            events.PARTICLE_MACRO,
            `0.5, -0.5, eatfood 10 ${holder.eid}`,
            holder
          )
        );
        break;

      case 'w':
        holder.region.broadcast(
          new events.Event(
            events.DIRECT_ATTACK,
            `${x} ${y} ${this.slots[i]}`,
            holder
          )
        );
        break;

      case 'p':
        this.remove(i);

        const effectN = Math.random() * 4 | 0;
        const effectDuration = (Math.random() * 10 | 0) + 10;
        switch(effectN) {
          case 0:
            holder.setEffect('flip', effectDuration);
            break;
          case 1:
            holder.setEffect('invincible', effectDuration);
            break;
          case 2:
            holder.setEffect('blindness', effectDuration);
            break;
          case 3:
            holder.setEffect('drained', effectDuration);
            break;
        }
        holder.region.broadcast(
          new events.Event(
            events.SOUND,
            `potion${Math.random() * 2 | 0}:${x}:${y}`,
            null
          )
        );
        break;
    }
  }

  remove(i) {
    this.counts[i] -= 1;
    if (!this.counts[i]) {
      this.slots[i] = null;
    }
    this.consolidate();
    this.owner.updateInventory();
  }
  drop(dropper) {
    if (!this.slots[0]) {
      return;
    }

    console.log(`${dropper} dropping ${this.slots[0]}`);

    const region = dropper.region;
    const item = new ItemEntity(this.slots[0], dropper);
    this.counts[0] -= 1;

    if (!this.counts[0]) {
      this.slots[0] = null;
    }
    region.addEntity(item);
    this.consolidate();
    this.owner.updateInventory();
  }
};
